File: src/main/java/graphql/nadel/Nadel.java
package graphql.nadel;

import graphql.ExecutionInput;
import graphql.ExecutionResult;
import graphql.ExecutionResultImpl;
import graphql.ParseAndValidateResult;
import graphql.PublicApi;
import graphql.execution.AbortExecutionException;
import graphql.execution.ExecutionId;
import graphql.execution.ExecutionIdProvider;
import graphql.execution.instrumentation.DocumentAndVariables;
import graphql.execution.instrumentation.InstrumentationContext;
import graphql.execution.instrumentation.InstrumentationState;
import graphql.execution.preparsed.NoOpPreparsedDocumentProvider;
import graphql.execution.preparsed.PreparsedDocumentEntry;
import graphql.execution.preparsed.PreparsedDocumentProvider;
import graphql.language.Document;
import graphql.nadel.dsl.CommonDefinition;
import graphql.nadel.dsl.ServiceDefinition;
import graphql.nadel.dsl.StitchingDsl;
import graphql.nadel.engine.Execution;
import graphql.nadel.hooks.ServiceExecutionHooks;
import graphql.nadel.instrumentation.NadelInstrumentation;
import graphql.nadel.instrumentation.parameters.NadelInstrumentationCreateStateParameters;
import graphql.nadel.instrumentation.parameters.NadelInstrumentationQueryExecutionParameters;
import graphql.nadel.instrumentation.parameters.NadelNadelInstrumentationQueryValidationParameters;
import graphql.nadel.introspection.DefaultIntrospectionRunner;
import graphql.nadel.introspection.IntrospectionRunner;
import graphql.nadel.schema.NeverWiringFactory;
import graphql.nadel.schema.OverallSchemaGenerator;
import graphql.nadel.schema.SchemaTransformationHook;
import graphql.nadel.schema.UnderlyingSchemaGenerator;
import graphql.nadel.util.FpKit;
import graphql.nadel.util.LogKit;
import graphql.parser.InvalidSyntaxException;
import graphql.schema.GraphQLSchema;
import graphql.schema.idl.ScalarInfo;
import graphql.schema.idl.TypeDefinitionRegistry;
import graphql.schema.idl.WiringFactory;
import graphql.validation.ValidationError;
import graphql.validation.Validator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import java.util.function.UnaryOperator;

import static graphql.execution.instrumentation.DocumentAndVariables.newDocumentAndVariables;
import static graphql.nadel.util.Util.buildServiceRegistry;
import static java.util.Objects.requireNonNull;

@PublicApi
public class Nadel {

    private static final Logger logNotSafe = LogKit.getNotPrivacySafeLogger(Nadel.class);
    private static final Logger log = LoggerFactory.getLogger(Nadel.class);

    private final StitchingDsl stitchingDsl;
    private final ServiceExecutionFactory serviceExecutionFactory;
    private final NSDLParser NSDLParser = new NSDLParser();
    private final List<Service> services;
    private final GraphQLSchema overallSchema;
    private final NadelInstrumentation instrumentation;
    private final ServiceExecutionHooks serviceExecutionHooks;
    private final PreparsedDocumentProvider preparsedDocumentProvider;
    private final ExecutionIdProvider executionIdProvider;
    private final IntrospectionRunner introspectionRunner;
    private final DefinitionRegistry commonTypes;
    private final WiringFactory overallWiringFactory;
    private final WiringFactory underlyingWiringFactory;
    private final SchemaTransformationHook schemaTransformationHook;
    private final OverallSchemaGenerator overallSchemaGenerator = new OverallSchemaGenerator();

    private Nadel(Reader nsdl,
                  ServiceExecutionFactory serviceExecutionFactory,
                  NadelInstrumentation instrumentation,
                  PreparsedDocumentProvider preparsedDocumentProvider,
                  ExecutionIdProvider executionIdProvider,
                  IntrospectionRunner introspectionRunner,
                  ServiceExecutionHooks serviceExecutionHooks,
                  WiringFactory overallWiringFactory,
                  WiringFactory underlyingWiringFactory,
                  SchemaTransformationHook schemaTransformationHook) {
        this.serviceExecutionFactory = serviceExecutionFactory;
        this.instrumentation = instrumentation;
        this.serviceExecutionHooks = serviceExecutionHooks;
        this.preparsedDocumentProvider = preparsedDocumentProvider;
        this.executionIdProvider = executionIdProvider;
        this.schemaTransformationHook = schemaTransformationHook;

        this.stitchingDsl = this.NSDLParser.parseDSL(nsdl);
        this.introspectionRunner = introspectionRunner;
        this.overallWiringFactory = overallWiringFactory;
        this.underlyingWiringFactory = underlyingWiringFactory;
        this.services = createServices();
        this.commonTypes = createCommonTypes();
        this.overallSchema = createOverallSchema();
    }

    private DefinitionRegistry createCommonTypes() {
        CommonDefinition commonDefinition = stitchingDsl.getCommonDefinition();
        return buildServiceRegistry(commonDefinition);
    }

    private List<Service> createServices() {
        List<ServiceDefinition> serviceDefinitions = stitchingDsl.getServiceDefinitions();

        UnderlyingSchemaGenerator underlyingSchemaGenerator = new UnderlyingSchemaGenerator();

        List<Service> serviceList = new ArrayList<>();
        for (ServiceDefinition serviceDefinition : serviceDefinitions) {
            String serviceName = serviceDefinition.getName();
            ServiceExecution serviceExecution = this.serviceExecutionFactory.getServiceExecution(serviceName);
            TypeDefinitionRegistry underlyingTypeDefinitions = this.serviceExecutionFactory.getUnderlyingTypeDefinitions(serviceName);

            GraphQLSchema underlyingSchema = underlyingSchemaGenerator
                    .buildUnderlyingSchema(serviceName, underlyingTypeDefinitions, underlyingWiringFactory);
            DefinitionRegistry definitionRegistry = buildServiceRegistry(serviceDefinition);

            Service service = new Service(serviceName, underlyingSchema, serviceExecution, serviceDefinition, definitionRegistry);
            serviceList.add(service);
        }

        return serviceList;

    }

    private GraphQLSchema createOverallSchema() {
        List<DefinitionRegistry> registries = FpKit.map(this.services, Service::getDefinitionRegistry);
        GraphQLSchema schema = overallSchemaGenerator.buildOverallSchema(registries, commonTypes, overallWiringFactory);

        GraphQLSchema newSchema = schemaTransformationHook.apply(schema);

        //
        // make sure that the overall schema has the standard scalars in it since he underlying may use them EVEN if the overall does not
        // make direct use of them, we still have to map between them
        return newSchema.transform(builder -> ScalarInfo.GRAPHQL_SPECIFICATION_SCALARS.forEach(builder::additionalType));
    }

    public List<Service> getServices() {
        return services;
    }

    public GraphQLSchema getOverallSchema() {
        return overallSchema;
    }

    public CompletableFuture<ExecutionResult> execute(NadelExecutionInput.Builder nadelExecutionInput) {
        return execute(nadelExecutionInput.build());
    }

    public CompletableFuture<ExecutionResult> execute(UnaryOperator<NadelExecutionInput.Builder> builderFunction) {
        return execute(builderFunction.apply(NadelExecutionInput.newNadelExecutionInput()).build());
    }

    public CompletableFuture<ExecutionResult> execute(NadelExecutionInput nadelExecutionInput) {
        long startTime = System.currentTimeMillis();
        ExecutionInput executionInput = ExecutionInput.newExecutionInput()
                .query(nadelExecutionInput.getQuery())
                .operationName(nadelExecutionInput.getOperationName())
                .context(nadelExecutionInput.getContext())
                .variables(nadelExecutionInput.getVariables())
                .executionId(nadelExecutionInput.getExecutionId())
                .build();

        NadelExecutionParams nadelExecutionParams = new NadelExecutionParams(nadelExecutionInput.getArtificialFieldsUUID());

        InstrumentationState instrumentationState = instrumentation.createState(new NadelInstrumentationCreateStateParameters(overallSchema, executionInput));
        NadelInstrumentationQueryExecutionParameters instrumentationParameters = new NadelInstrumentationQueryExecutionParameters(executionInput, overallSchema, instrumentationState);
        try {
            logNotSafe.debug("Executing request. operation name: '{}'. query: '{}'. variables '{}'", executionInput.getOperationName(), executionInput.getQuery(), executionInput.getVariables());

            NadelInstrumentationQueryExecutionParameters inputInstrumentationParameters = new NadelInstrumentationQueryExecutionParameters(executionInput, overallSchema, instrumentationState);
            executionInput = instrumentation.instrumentExecutionInput(executionInput, inputInstrumentationParameters);

            InstrumentationContext<ExecutionResult> executionInstrumentation = instrumentation.beginQueryExecution(instrumentationParameters);

            CompletableFuture<ExecutionResult> executionResult = parseValidateAndExecute(executionInput, overallSchema, instrumentationState, nadelExecutionParams);
            //
            // finish up instrumentation
            executionResult = executionResult.whenComplete(executionInstrumentation::onCompleted);
            //
            // allow instrumentation to tweak the result
            executionResult = executionResult.thenCompose(result -> instrumentation.instrumentExecutionResult(result, instrumentationParameters));
            return executionResult.whenComplete((executionResult1, throwable) -> {
                long elapsedTime = System.currentTimeMillis() - startTime;
                log.debug("Finished execution in {} ms, executionId: {}", elapsedTime, nadelExecutionInput.getExecutionId());
            });
        } catch (AbortExecutionException abortException) {
            return instrumentation.instrumentExecutionResult(abortException.toExecutionResult(), instrumentationParameters);
        }
    }

    private CompletableFuture<ExecutionResult> parseValidateAndExecute(ExecutionInput executionInput,
                                                                       GraphQLSchema graphQLSchema,
                                                                       InstrumentationState instrumentationState,
                                                                       NadelExecutionParams nadelExecutionParams) {
        AtomicReference<ExecutionInput> executionInputRef = new AtomicReference<>(executionInput);
        Function<ExecutionInput, PreparsedDocumentEntry> computeFunction = transformedInput -> {
            // if they change the original query in the pre-parser, then we want to see it downstream from then on
            executionInputRef.set(transformedInput);
            return parseAndValidate(executionInputRef, graphQLSchema, instrumentationState);
        };
        PreparsedDocumentEntry preparsedDoc = preparsedDocumentProvider.getDocument(executionInput, computeFunction);
        if (preparsedDoc.hasErrors()) {
            return CompletableFuture.completedFuture(new ExecutionResultImpl(preparsedDoc.getErrors()));
        }
        return executeImpl(executionInputRef.get(), preparsedDoc.getDocument(), instrumentationState, nadelExecutionParams);
    }

    private PreparsedDocumentEntry parseAndValidate(AtomicReference<ExecutionInput> executionInputRef, GraphQLSchema graphQLSchema, InstrumentationState instrumentationState) {

        ExecutionInput executionInput = executionInputRef.get();
        String query = executionInput.getQuery();

        logNotSafe.debug("Parsing query: '{}'...", query);
        ParseAndValidateResult parseResult = parse(executionInput, graphQLSchema, instrumentationState);
        if (parseResult.isFailure()) {
            logNotSafe.warn("Query failed to parse : '{}'", executionInput.getQuery());
            return new PreparsedDocumentEntry(parseResult.getSyntaxException().toInvalidSyntaxError());
        } else {
            final Document document = parseResult.getDocument();
            // they may have changed the document and the variables via instrumentation so update the reference to it
            executionInput = executionInput.transform(builder -> builder.variables(parseResult.getVariables()));
            executionInputRef.set(executionInput);

            logNotSafe.debug("Validating query: '{}'", query);
            final List<ValidationError> errors = validate(executionInput, document, graphQLSchema, instrumentationState);
            if (!errors.isEmpty()) {
                logNotSafe.warn("Query failed to validate : '{}' because of {} ", query, errors);
                return new PreparsedDocumentEntry(errors);
            }

            return new PreparsedDocumentEntry(document);
        }
    }

    private ParseAndValidateResult parse(ExecutionInput executionInput, GraphQLSchema graphQLSchema, InstrumentationState instrumentationState) {
        NadelInstrumentationQueryExecutionParameters parameters = new NadelInstrumentationQueryExecutionParameters(executionInput, graphQLSchema, instrumentationState);
        InstrumentationContext<Document> parseInstrumentation = instrumentation.beginParse(parameters);

        Document document;
        DocumentAndVariables documentAndVariables;
        try {
            document = new NadelGraphQLParser().parseDocument(executionInput.getQuery());
            documentAndVariables = newDocumentAndVariables()
                    .document(document).variables(executionInput.getVariables()).build();
            documentAndVariables = instrumentation.instrumentDocumentAndVariables(documentAndVariables, parameters);
        } catch (InvalidSyntaxException e) {
            parseInstrumentation.onCompleted(null, e);
            return ParseAndValidateResult.newResult().syntaxException(e).build();
        }

        parseInstrumentation.onCompleted(documentAndVariables.getDocument(), null);
        return ParseAndValidateResult.newResult().document(documentAndVariables.getDocument()).variables(documentAndVariables.getVariables()).build();
    }

    private List<ValidationError> validate(ExecutionInput executionInput, Document document, GraphQLSchema graphQLSchema, InstrumentationState instrumentationState) {
        InstrumentationContext<List<ValidationError>> validationCtx = instrumentation.beginValidation(new NadelNadelInstrumentationQueryValidationParameters(executionInput, document, graphQLSchema, instrumentationState));

        Validator validator = new Validator();
        List<ValidationError> validationErrors = validator.validateDocument(graphQLSchema, document);

        validationCtx.onCompleted(validationErrors, null);
        return validationErrors;
    }

    private CompletableFuture<ExecutionResult> executeImpl(ExecutionInput executionInput,
                                                           Document document,
                                                           InstrumentationState instrumentationState,
                                                           NadelExecutionParams nadelExecutionParams) {

        String query = executionInput.getQuery();
        String operationName = executionInput.getOperationName();
        Object context = executionInput.getContext();

        ExecutionId executionId = executionInput.getExecutionId();
        if (executionId == null) {
            executionId = executionIdProvider.provide(query, operationName, context);
        }

        if (executionInput.getContext() instanceof BenchmarkContext) {
            BenchmarkContext.ExecutionArgs executionArgs = ((BenchmarkContext) executionInput.getContext()).executionArgs;
            executionArgs.services = getServices();
            executionArgs.overallSchema = overallSchema;
            executionArgs.instrumentation = instrumentation;
            executionArgs.introspectionRunner = introspectionRunner;
            executionArgs.serviceExecutionHooks = serviceExecutionHooks;
            executionArgs.context = executionInput.getContext();
            executionArgs.executionInput = executionInput;
            executionArgs.document = document;
            executionArgs.executionId = executionId;
            executionArgs.instrumentationState = instrumentationState;
            executionArgs.nadelExecutionParams = nadelExecutionParams;
        }

        Execution execution = new Execution(getServices(), overallSchema, instrumentation, introspectionRunner, serviceExecutionHooks, executionInput.getContext());

        return execution.execute(executionInput, document, executionId, instrumentationState, nadelExecutionParams);
    }

    /**
     * @return a builder of Nadel objects
     */
    public static Builder newNadel() {
        return new Builder();
    }

    public static class Builder {
        private Reader nsdl;
        private ServiceExecutionFactory serviceExecutionFactory;
        private NadelInstrumentation instrumentation = new NadelInstrumentation() {
        };
        private ServiceExecutionHooks serviceExecutionHooks = new ServiceExecutionHooks() {
        };
        private PreparsedDocumentProvider preparsedDocumentProvider = NoOpPreparsedDocumentProvider.INSTANCE;
        private ExecutionIdProvider executionIdProvider = ExecutionIdProvider.DEFAULT_EXECUTION_ID_PROVIDER;
        private IntrospectionRunner introspectionRunner = new DefaultIntrospectionRunner();
        private WiringFactory overallWiringFactory = new NeverWiringFactory();
        private WiringFactory underlyingWiringFactory = new NeverWiringFactory();
        private SchemaTransformationHook schemaTransformationHook = SchemaTransformationHook.IDENTITY;


        public Builder dsl(Reader nsdl) {
            this.nsdl = requireNonNull(nsdl);
            return this;
        }

        public Builder dsl(String nsdl) {
            return dsl(new StringReader(requireNonNull(nsdl)));
        }

        public Builder serviceExecutionFactory(ServiceExecutionFactory serviceExecutionFactory) {
            this.serviceExecutionFactory = serviceExecutionFactory;
            return this;
        }

        public Builder instrumentation(NadelInstrumentation instrumentation) {
            this.instrumentation = requireNonNull(instrumentation);
            return this;
        }

        public Builder preparsedDocumentProvider(PreparsedDocumentProvider preparsedDocumentProvider) {
            this.preparsedDocumentProvider = requireNonNull(preparsedDocumentProvider);
            return this;
        }

        public Builder executionIdProvider(ExecutionIdProvider executionIdProvider) {
            this.executionIdProvider = requireNonNull(executionIdProvider);
            return this;
        }

        public Builder introspectionRunner(IntrospectionRunner introspectionRunner) {
            this.introspectionRunner = requireNonNull(introspectionRunner);
            return this;
        }

        public Builder serviceExecutionHooks(ServiceExecutionHooks serviceExecutionHooks) {
            this.serviceExecutionHooks = requireNonNull(serviceExecutionHooks);
            return this;
        }

        public Builder overallWiringFactory(WiringFactory wiringFactory) {
            this.overallWiringFactory = requireNonNull(wiringFactory);
            return this;
        }

        public Builder underlyingWiringFactory(WiringFactory wiringFactory) {
            this.underlyingWiringFactory = requireNonNull(wiringFactory);
            return this;
        }

        public Builder schemaTransformationHook(SchemaTransformationHook hook) {
            this.schemaTransformationHook = hook;
            return this;
        }

        public Nadel build() {
            return new Nadel(
                    nsdl,
                    serviceExecutionFactory,
                    instrumentation,
                    preparsedDocumentProvider,
                    executionIdProvider,
                    introspectionRunner,
                    serviceExecutionHooks,
                    overallWiringFactory,
                    underlyingWiringFactory,
                    schemaTransformationHook);
        }
    }
}

File: lib/src/main/java/graphql/nadel/Nadel.kt
package graphql.nadel

import graphql.ExecutionInput
import graphql.ExecutionInput.newExecutionInput
import graphql.ExecutionResult
import graphql.ExecutionResultImpl
import graphql.ParseAndValidateResult
import graphql.execution.AbortExecutionException
import graphql.execution.ExecutionIdProvider
import graphql.execution.instrumentation.DocumentAndVariables
import graphql.execution.instrumentation.InstrumentationState
import graphql.execution.preparsed.NoOpPreparsedDocumentProvider
import graphql.execution.preparsed.PreparsedDocumentEntry
import graphql.execution.preparsed.PreparsedDocumentProvider
import graphql.language.Document
import graphql.nadel.engine.blueprint.NadelDefaultIntrospectionRunner
import graphql.nadel.engine.blueprint.NadelIntrospectionRunnerFactory
import graphql.nadel.engine.transform.NadelTransform
import graphql.nadel.hooks.NadelExecutionHooks
import graphql.nadel.instrumentation.NadelInstrumentation
import graphql.nadel.instrumentation.parameters.NadelInstrumentationCreateStateParameters
import graphql.nadel.instrumentation.parameters.NadelInstrumentationQueryExecutionParameters
import graphql.nadel.instrumentation.parameters.NadelInstrumentationQueryValidationParameters
import graphql.nadel.schema.QuerySchemaGenerator
import graphql.nadel.schema.SchemaTransformationHook
import graphql.nadel.util.LogKit
import graphql.parser.InvalidSyntaxException
import graphql.parser.Parser
import graphql.schema.GraphQLSchema
import graphql.schema.idl.TypeDefinitionRegistry
import graphql.schema.idl.WiringFactory
import graphql.validation.ValidationError
import graphql.validation.Validator
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.future.await
import kotlinx.coroutines.future.future
import kotlinx.coroutines.launch
import org.slf4j.Logger
import java.io.Reader
import java.io.StringReader
import java.util.Locale
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.concurrent.atomic.AtomicReference

class Nadel private constructor(
    private val engine: NextgenEngine,
    val services: List<Service>,
    val engineSchema: GraphQLSchema,
    val querySchema: GraphQLSchema,
    private val instrumentation: NadelInstrumentation,
    private val preparsedDocumentProvider: PreparsedDocumentProvider,
) {
    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    @Deprecated(message = "Use executeAsync", replaceWith = ReplaceWith("this.executeAsync(executionInput)"))
    @JvmName("execute") // For binary compat
    fun executeJava(executionInput: NadelExecutionInput): CompletableFuture<ExecutionResult> {
        return coroutineScope.future {
            execute(executionInput)
        }
    }

    fun executeAsync(executionInput: NadelExecutionInput): CompletableFuture<ExecutionResult> {
        return coroutineScope.future {
            execute(executionInput)
        }
    }

    @JvmName("executeSuspending")
    suspend fun execute(nadelExecutionInput: NadelExecutionInput): ExecutionResult {
        val executionInput: ExecutionInput = newExecutionInput()
            .query(nadelExecutionInput.query)
            .operationName(nadelExecutionInput.operationName)
            .context(nadelExecutionInput.context)
            .graphQLContext { builder ->
                // we need to transfer over all the keys - ExecutionInput does not allow for setting it direct
                nadelExecutionInput.graphqlContext.stream().forEach {
                    builder.put(it.key, it.value)
                }
            }
            .variables(nadelExecutionInput.variables)
            .extensions(nadelExecutionInput.extensions)
            .executionId(nadelExecutionInput.executionId)
            .build()

        val hints = nadelExecutionInput.nadelExecutionHints
        val instrumentationState = instrumentation.createState(
            NadelInstrumentationCreateStateParameters(querySchema, executionInput),
        )
        val instrumentationParameters = NadelInstrumentationQueryExecutionParameters(
            executionInput, querySchema, instrumentationState
        )

        return try {
            val executionInstrumentation = instrumentation.beginQueryExecution(instrumentationParameters)

            val result = try {
                parseValidateAndExecute(executionInput, instrumentationState, hints)
                    .also {
                        executionInstrumentation.onCompleted(it, null)
                    }
            } catch (e: Exception) {
                executionInstrumentation.onCompleted(null, e)

                val cause = e.cause
                if (e is AbortExecutionException) {
                    e.toExecutionResult()
                } else if (e is CompletionException && cause is AbortExecutionException) {
                    cause.toExecutionResult()
                } else {
                    throw e
                }
            }

            instrumentation
                .instrumentExecutionResult(result, instrumentationParameters)
                .await()
        } catch (e: AbortExecutionException) {
            instrumentation
                .instrumentExecutionResult(e.toExecutionResult(), instrumentationParameters)
                .await()
        }
    }

    fun close() {
        // Closes the scope after letting in flight requests go through
        coroutineScope.launch {
            delay(60_000) // Wait a minute
            coroutineScope.cancel()
        }
    }

    private suspend fun parseValidateAndExecute(
        executionInput: ExecutionInput,
        instrumentationState: InstrumentationState?,
        hints: NadelExecutionHints,
    ): ExecutionResult {
        // todo: I'm pretty sure this is never changed, but let's circle back to that in another PR to reduce changelog here
        val executionInputRef = AtomicReference(executionInput)

        val result = preparsedDocumentProvider
            .getDocumentAsync(executionInput) { transformedInput ->
                // If they change the original query in the pre-parser, then we want to see it downstream from then on
                executionInputRef.set(transformedInput)
                parseAndValidate(executionInputRef, instrumentationState)
            }
            .await()

        return if (result.hasErrors()) {
            ExecutionResultImpl(result.errors)
        } else {
            coroutineScope {
                engine.execute(
                    coroutineScope = this,
                    executionInput = executionInputRef.get()!!,
                    queryDocument = result.document,
                    instrumentationState = instrumentationState,
                    executionHints = hints,
                )
            }
        }
    }

    private fun parseAndValidate(
        executionInputRef: AtomicReference<ExecutionInput>,
        instrumentationState: InstrumentationState?,
    ): PreparsedDocumentEntry {
        var executionInput = executionInputRef.get()!!

        val query = executionInput.query
        logNotSafe.debug("Parsing query: '{}'...", query)
        val parseResult = parse(executionInput, instrumentationState)

        return if (parseResult.isFailure) {
            logNotSafe.warn("Query failed to parse : '{}'", executionInput.query)
            PreparsedDocumentEntry(parseResult.syntaxException.toInvalidSyntaxError())
        } else {
            val document = parseResult.document

            // they may have changed the document and the variables via instrumentation so update the reference to it
            executionInput = executionInput.transform { builder: ExecutionInput.Builder ->
                builder.variables(parseResult.variables)
            }
            executionInputRef.set(executionInput)

            logNotSafe.debug("Validating query: '{}'", query)
            val errors = validate(executionInput, document, instrumentationState)

            if (errors.isNotEmpty()) {
                logNotSafe.warn("Query failed to validate : '{}' because of {} ", query, errors)
                PreparsedDocumentEntry(errors)
            } else {
                PreparsedDocumentEntry(document)
            }
        }
    }

    private fun parse(
        executionInput: ExecutionInput,
        instrumentationState: InstrumentationState?,
    ): ParseAndValidateResult {
        val parameters = NadelInstrumentationQueryExecutionParameters(
            executionInput,
            querySchema,
            instrumentationState
        )

        val parseInstrumentation = instrumentation.beginParse(parameters)
        val document: Document
        val documentAndVariables: DocumentAndVariables

        try {
            document = Parser().parseDocument(executionInput.query)
            documentAndVariables = DocumentAndVariables.newDocumentAndVariables()
                .document(document)
                .variables(executionInput.variables)
                .build()
        } catch (e: InvalidSyntaxException) {
            parseInstrumentation.onCompleted(null, e)
            return ParseAndValidateResult.newResult().syntaxException(e).build()
        }

        parseInstrumentation.onCompleted(documentAndVariables.document, null)

        return ParseAndValidateResult
            .newResult()
            .document(documentAndVariables.document)
            .variables(documentAndVariables.variables)
            .build()
    }

    private fun validate(
        executionInput: ExecutionInput,
        document: Document,
        instrumentationState: InstrumentationState?,
    ): MutableList<ValidationError> {
        val validationCtx = instrumentation.beginValidation(
            NadelInstrumentationQueryValidationParameters(
                executionInput = executionInput,
                document = document,
                schema = querySchema,
                instrumentationState = instrumentationState,
                context = executionInput.context,
            ),
        )
        val validator = Validator()
        val validationErrors = validator.validateDocument(querySchema, document, Locale.getDefault())
        validationCtx.onCompleted(validationErrors, null)
        return validationErrors
    }

    class Builder {
        private var instrumentation: NadelInstrumentation = object : NadelInstrumentation {}
        private var executionHooks: NadelExecutionHooks = object : NadelExecutionHooks {}
        private var preparsedDocumentProvider: PreparsedDocumentProvider = NoOpPreparsedDocumentProvider.INSTANCE
        private var executionIdProvider = ExecutionIdProvider.DEFAULT_EXECUTION_ID_PROVIDER
        private var transforms = emptyList<NadelTransform<out Any>>()
        private var introspectionRunnerFactory = NadelIntrospectionRunnerFactory(::NadelDefaultIntrospectionRunner)

        private var schemas: NadelSchemas? = null
        private var schemaBuilder = NadelSchemas.Builder()

        private var maxQueryDepth = Integer.MAX_VALUE
        private var maxFieldCount = Integer.MAX_VALUE

        fun schemas(schemas: NadelSchemas): Builder {
            this.schemas = schemas
            return this
        }

        fun overallSchema(serviceName: String, nsdl: Reader): Builder {
            schemaBuilder.overallSchema(serviceName, nsdl)
            return this
        }

        fun overallSchema(serviceName: String, nsdl: String): Builder {
            schemaBuilder.overallSchema(serviceName, nsdl)
            return this
        }

        @JvmName("overallSchemasReader")
        fun overallSchemas(overallSchemas: Map<String, Reader>): Builder {
            schemaBuilder.overallSchemas(overallSchemas)
            return this
        }

        @JvmName("overallSchemasString")
        fun overallSchemas(overallSchemas: Map<String, String>): Builder {
            schemaBuilder.overallSchemas(overallSchemas)
            return this
        }

        fun underlyingSchema(serviceName: String, nsdl: Reader): Builder {
            schemaBuilder.underlyingSchema(serviceName, nsdl)
            return this
        }

        fun underlyingSchema(serviceName: String, nsdl: String): Builder {
            schemaBuilder.underlyingSchema(serviceName, nsdl)
            return this
        }

        fun underlyingSchema(serviceName: String, schema: TypeDefinitionRegistry): Builder {
            schemaBuilder.underlyingSchema(serviceName, schema)
            return this
        }

        @JvmName("underlyingSchemasReader")
        fun underlyingSchemas(underlyingSchemas: Map<String, Reader>): Builder {
            schemaBuilder.underlyingSchemas(underlyingSchemas)
            return this
        }

        @JvmName("underlyingSchemasString")
        fun underlyingSchemas(underlyingSchemas: Map<String, String>): Builder {
            schemaBuilder.underlyingSchemas(underlyingSchemas)
            return this
        }

        @JvmName("underlyingTypeDefs")
        fun underlyingSchemas(value: Map<String, TypeDefinitionRegistry>): Builder = also {
            schemaBuilder.underlyingSchemas(value)
            return this
        }

        fun overallWiringFactory(wiringFactory: WiringFactory): Builder {
            schemaBuilder.overallWiringFactory(wiringFactory)
            return this
        }

        fun underlyingWiringFactory(wiringFactory: WiringFactory): Builder {
            schemaBuilder.underlyingWiringFactory(wiringFactory)
            return this
        }

        fun schemaTransformationHook(hook: SchemaTransformationHook): Builder {
            schemaBuilder.schemaTransformationHook(hook)
            return this
        }

        fun serviceExecutionFactory(serviceExecutionFactory: ServiceExecutionFactory): Builder {
            schemaBuilder.serviceExecutionFactory(serviceExecutionFactory)
            return this
        }

        fun instrumentation(instrumentation: NadelInstrumentation): Builder {
            this.instrumentation = instrumentation
            return this
        }

        fun preparsedDocumentProvider(preparsedDocumentProvider: PreparsedDocumentProvider): Builder {
            this.preparsedDocumentProvider = preparsedDocumentProvider
            return this
        }

        fun executionIdProvider(executionIdProvider: ExecutionIdProvider): Builder {
            this.executionIdProvider = executionIdProvider
            return this
        }

        fun executionHooks(executionHooks: NadelExecutionHooks): Builder {
            this.executionHooks = executionHooks
            return this
        }

        fun transforms(transforms: List<NadelTransform<out Any>>): Builder {
            this.transforms = transforms
            return this
        }

        fun introspectionRunnerFactory(introspectionRunnerFactory: NadelIntrospectionRunnerFactory): Builder {
            this.introspectionRunnerFactory = introspectionRunnerFactory
            return this
        }

        fun maxQueryDepth(maxQueryDepth: Int): Builder {
            this.maxQueryDepth = maxQueryDepth
            return this
        }

        fun maxFieldCount(maxFieldCount: Int): Builder {
            this.maxFieldCount = maxFieldCount
            return this
        }

        fun build(): Nadel {
            val (engineSchema, services) = schemas ?: schemaBuilder.build()

            val querySchema = QuerySchemaGenerator.generateQuerySchema(engineSchema)

            return Nadel(
                engine = NextgenEngine(
                    engineSchema = engineSchema,
                    querySchema = querySchema,
                    instrumentation = instrumentation,
                    executionHooks = executionHooks,
                    executionIdProvider = executionIdProvider,
                    services = services,
                    transforms = transforms,
                    introspectionRunnerFactory = introspectionRunnerFactory,
                    maxQueryDepth = maxQueryDepth,
                    maxFieldCount = maxFieldCount,
                ),
                services = services,
                engineSchema = engineSchema,
                querySchema = querySchema,
                instrumentation = instrumentation,
                preparsedDocumentProvider = preparsedDocumentProvider,
            )
        }

        @Deprecated("Use overallSchema instead", replaceWith = ReplaceWith("overallSchema(serviceName, nsdl)"))
        fun dsl(serviceName: String, nsdl: Reader): Builder {
            overallSchema(serviceName, nsdl)
            return this
        }

        @Deprecated("Use overallSchema instead", replaceWith = ReplaceWith("overallSchema(serviceName, nsdl)"))
        fun dsl(serviceName: String, nsdl: String): Builder {
            return dsl(serviceName, StringReader(nsdl))
        }

        @Deprecated("Use overallSchemas instead", replaceWith = ReplaceWith("overallSchemas(serviceDSLs)"))
        fun dsl(serviceDSLs: Map<String, String>): Builder {
            return serviceDSLs(serviceDSLs.mapValues { (k, v) -> StringReader(v) })
        }

        @Deprecated("Use overallSchemas instead", replaceWith = ReplaceWith("overallSchemas(serviceDSLs)"))
        fun serviceDSLs(serviceDSLs: Map<String, Reader>): Builder {
            schemaBuilder.overallSchemas(serviceDSLs)
            return this
        }
    }

    companion object {
        private val logNotSafe: Logger = LogKit.getNotPrivacySafeLogger<Nadel>()
        private val log: Logger = LogKit.getLogger<Nadel>()

        /**
         * @return a builder of Nadel objects
         */
        @JvmStatic
        fun newNadel(): Builder {
            return Builder()
        }
    }
}
